src/ray/raylet/node_manager.cc

```diff
@@ -949,8 +951,18 @@ void NodeManager::ProcessRegisterClientRequestMessage(
   auto message = flatbuffers::GetRoot<protocol::RegisterClientRequest>(message_data);
   Language language = static_cast<Language>(message->language());
   WorkerID worker_id = from_flatbuf<WorkerID>(*message->worker_id());
-  auto worker = std::make_shared<Worker>(worker_id, message->worker_pid(), language,
-                                         message->port(), client, client_call_manager_);
+  WorkerProcessHandle proc =
+      worker_pool_.FindStartingWorkerByProcessId(language, message->worker_pid());
+  if (!proc) {
+    // Unknown worker! We'll still use it and let the pool deal with it on registration.
```

Comment:
When would this happen?

Reply:
Ah, I forgot to take this out. Originally I was trying to delegate to the handler [here](https://github.com/mehrdadn/ray/blob/boost-process/src/ray/raylet/worker_pool.cc#L227), but @edoakes pointed out that that code is already exiting, so we should probably do that here too.

However, now I'm wondering something: should we worry about what happens if the raylet dies and a new one gets spawned (which wouldn't be the parent of the original workers anymore)? I'm guessing it's not currently an issue since we just terminate workers when the raylet dies, so maybe we should leave worrying about that to future PRs that might try to handle it differently. Just worth keeping in mind that it's going to be a bit incompatible with Linux's process model, since there isn't an elegant way to wait on others' children.

Reply:
Actually there is a case where this could happen: in some unit tests we reuse the same raylet socket path across different tests, and if some workers start slowly and haven't registered themselves with raylet before this test case finishes, they could end up registering with the new raylet when the next test case starts. 

So we cannot RAY_CHECK in this case, just ignore the register request and print a WARNING message should be sufficient.

Reply:
Wouldn't(/shouldn't) those workers be killed before the next test starts?

Reply:
Raylet gets killed after a test case finishes, thus it doesn't have chance to clean up these workers. For other workers that have registered, they have connections to raylet, and would die after they detect the connection failure.

Reply:
If I understand what you're saying correctly, I think the proper solution here is for the workers to avoid needing to connect to a UNIX socket altogether, but to rather for the raylet to just duplicate a pipe into every worker process so that the worker already has the necessary connection on startup. That's probably best left for another PR, but I'm happy to look at it sometime later.

However, are you saying the killing _always_ happens, or that it merely _could_ happen if e.g. the test times out? Because I thought the raylet should get terminated gracefully (`SIGTERM`) and would only be killed forcefully if that operation times out. But in that case, the test would fail anyway, so there wouldn't be an issue, right?

Reply:
Even if we only do force kill when the test times out, it should only affect the problematic test case. The issue is it would fail the next test case because of the added RAY_CHECK.

Reply:
Ohh I understand now, okay thanks! I just pushed a commit to ignore unknown requests.